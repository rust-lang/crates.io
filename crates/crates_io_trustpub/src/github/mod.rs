mod claims;
#[cfg(any(test, feature = "test-helpers"))]
pub mod test_helpers;
pub mod validation;
mod workflows;

pub use claims::GitHubClaims;

pub const GITHUB_ISSUER_URL: &str = "https://token.actions.githubusercontent.com";
