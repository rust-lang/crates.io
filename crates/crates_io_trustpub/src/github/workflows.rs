use std::sync::LazyLock;

/// Extracts the workflow filename from a GitHub workflow reference.
///
/// In other words, it turns e.g. `rust-lang/regex/.github/workflows/ci.yml@refs/heads/main`
/// into `ci.yml`, or `None` if the reference is in an unexpected format.
#[allow(unused)]
pub(crate) fn extract_workflow_filename(workflow_ref: &str) -> Option<&str> {
    static WORKFLOW_REF_RE: LazyLock<regex::Regex> =
        LazyLock::new(|| regex::Regex::new(r"([^/]+\.(yml|yaml))(@.+)").unwrap());

    WORKFLOW_REF_RE
        .captures(workflow_ref)
        .and_then(|caps| caps.get(1))
        .map(|m| m.as_str())
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_extract_workflow_filename() {
        let test_cases = [
            // Well-formed workflow refs, including exceedingly obnoxious ones
            // with `@` or extra suffixes or `git` refs that look like workflows.
            (
                "foo/bar/.github/workflows/basic.yml@refs/heads/main",
                Some("basic.yml"),
            ),
            (
                "foo/bar/.github/workflows/basic.yaml@refs/heads/main",
                Some("basic.yaml"),
            ),
            (
                "foo/bar/.github/workflows/has-dash.yml@refs/heads/main",
                Some("has-dash.yml"),
            ),
            (
                "foo/bar/.github/workflows/has--dashes.yml@refs/heads/main",
                Some("has--dashes.yml"),
            ),
            (
                "foo/bar/.github/workflows/has--dashes-.yml@refs/heads/main",
                Some("has--dashes-.yml"),
            ),
            (
                "foo/bar/.github/workflows/has.period.yml@refs/heads/main",
                Some("has.period.yml"),
            ),
            (
                "foo/bar/.github/workflows/has..periods.yml@refs/heads/main",
                Some("has..periods.yml"),
            ),
            (
                "foo/bar/.github/workflows/has..periods..yml@refs/heads/main",
                Some("has..periods..yml"),
            ),
            (
                "foo/bar/.github/workflows/has_underscore.yml@refs/heads/main",
                Some("has_underscore.yml"),
            ),
            (
                "foo/bar/.github/workflows/nested@evil.yml@refs/heads/main",
                Some("nested@evil.yml"),
            ),
            (
                "foo/bar/.github/workflows/nested.yml@evil.yml@refs/heads/main",
                Some("nested.yml@evil.yml"),
            ),
            (
                "foo/bar/.github/workflows/extra@nested.yml@evil.yml@refs/heads/main",
                Some("extra@nested.yml@evil.yml"),
            ),
            (
                "foo/bar/.github/workflows/extra.yml@nested.yml@evil.yml@refs/heads/main",
                Some("extra.yml@nested.yml@evil.yml"),
            ),
            (
                "foo/bar/.github/workflows/basic.yml@refs/heads/misleading@branch.yml",
                Some("basic.yml"),
            ),
            (
                "foo/bar/.github/workflows/basic.yml@refs/heads/bad@branch@twomatches.yml",
                Some("basic.yml"),
            ),
            (
                "foo/bar/.github/workflows/foo.yml.yml@refs/heads/main",
                Some("foo.yml.yml"),
            ),
            (
                "foo/bar/.github/workflows/foo.yml.foo.yml@refs/heads/main",
                Some("foo.yml.foo.yml"),
            ),
            // Malformed workflow refs.
            (
                "foo/bar/.github/workflows/basic.wrongsuffix@refs/heads/main",
                None,
            ),
            ("foo/bar/.github/workflows/@refs/heads/main", None),
            ("foo/bar/.github/workflows/nosuffix@refs/heads/main", None),
            ("foo/bar/.github/workflows/.yml@refs/heads/main", None),
            ("foo/bar/.github/workflows/.yaml@refs/heads/main", None),
            ("foo/bar/.github/workflows/main.yml", None),
        ];

        for (input, expected) in test_cases {
            let result = super::extract_workflow_filename(input);
            assert_eq!(result, expected, "Input: {input}");
        }
    }
}
