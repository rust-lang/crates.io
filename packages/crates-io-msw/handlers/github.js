import { http, HttpResponse } from 'msw';

export default [
  http.head('https://raw.githubusercontent.com/:owner/:projec/HEAD/.github/workflows/:workflow_filename', () => {
    return new HttpResponse(null, { status: 404 });
  }),
];
