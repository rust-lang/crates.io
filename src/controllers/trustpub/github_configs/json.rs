use axum::Json;
use axum::extract::FromRequest;
use chrono::{DateTime, Utc};

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct GitHubConfig {
    #[schema(example = 42)]
    pub id: i32,
    #[schema(example = "regex")]
    #[serde(rename = "crate")]
    pub krate: String,
    #[schema(example = "rust-lang")]
    pub repository_owner: String,
    #[schema(example = 5430905)]
    pub repository_owner_id: i32,
    #[schema(example = "regex")]
    pub repository_name: String,
    #[schema(example = "ci.yml")]
    pub workflow_filename: String,
    #[schema(example = json!(null))]
    pub environment: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
pub struct NewGitHubConfig {
    #[schema(example = "regex")]
    #[serde(rename = "crate")]
    pub krate: String,
    #[schema(example = "rust-lang")]
    pub repository_owner: String,
    #[schema(example = "regex")]
    pub repository_name: String,
    #[schema(example = "ci.yml")]
    pub workflow_filename: String,
    #[schema(example = json!(null))]
    pub environment: Option<String>,
}

#[derive(Debug, Deserialize, FromRequest, utoipa::ToSchema)]
#[from_request(via(Json))]
pub struct CreateRequest {
    pub github_config: NewGitHubConfig,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct CreateResponse {
    pub github_config: GitHubConfig,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct ListResponse {
    pub github_configs: Vec<GitHubConfig>,
}
