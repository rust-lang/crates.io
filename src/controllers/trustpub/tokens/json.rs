use axum::Json;
use axum::extract::FromRequest;

#[derive(Debug, Deserialize, FromRequest, utoipa::ToSchema)]
#[from_request(via(Json))]
pub struct ExchangeRequest {
    pub jwt: String,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct ExchangeResponse {
    pub token: String,
}
