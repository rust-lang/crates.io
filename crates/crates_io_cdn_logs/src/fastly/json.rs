//! Imported from <https://github.com/rust-lang/simpleinfra/blob/4fb365809295de075d28d8b2d51f6f419537be7d/terragrunt/modules/crates-io/compute-static/src/log_line.rs>

use chrono::{DateTime, Utc};
use serde::Deserialize;
use std::borrow::Cow;

/// This struct corresponds to the JSON payload of a log line from
/// Fastly's CDN logs.
#[derive(Debug, Deserialize)]
#[serde(tag = "version")]
pub enum LogLine<'a> {
    #[serde(borrow, rename = "1")]
    V1(LogLineV1<'a>),
}

impl LogLine<'_> {
    pub fn date_time(&self) -> DateTime<Utc> {
        match self {
            LogLine::V1(line) => line.date_time,
        }
    }

    pub fn method(&self) -> &str {
        match self {
            LogLine::V1(line) => &line.method,
        }
    }

    pub fn url(&self) -> &str {
        match self {
            LogLine::V1(line) => &line.url,
        }
    }

    pub fn status(&self) -> u16 {
        match self {
            LogLine::V1(line) => line.status,
        }
    }
}

/// This struct corresponds to the `"version": "1"` variant of the [LogLine] enum.
///
/// Compared to the implementation in the [rust-lang/simpleinfra](https://github.com/rust-lang/simpleinfra/)
/// repository, there are a couple of differences:
///
/// - The `bytes` field is not included, because we don't need it.
/// - The `ip` field is not included, because we don't need it.
/// - The `method` and `status` fields are not optional, because we handle
///   parsing errors gracefully.
/// - The `date_time` field is using `chrono` like the rest of the
///   crates.io codebase.
/// - The `method` and `url` fields are using `Cow` to avoid
///   unnecessary allocations.
#[derive(Debug, Deserialize)]
pub struct LogLineV1<'a> {
    pub date_time: DateTime<Utc>,
    #[serde(borrow)]
    pub method: Cow<'a, str>,
    #[serde(borrow)]
    pub url: Cow<'a, str>,
    pub status: u16,
}

#[cfg(test)]
mod tests {
    use super::*;
    use claims::assert_ok;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse() {
        let input = r#"{"bytes":null,"date_time":"2024-01-16T16:03:04.44007323Z","ip":"45.79.107.220","method":"GET","status":403,"url":"https://static.staging.crates.io/?1705420437","version":"1"}"#;
        let output = assert_ok!(serde_json::from_str::<LogLine<'_>>(input));
        assert_debug_snapshot!(output, @r#"
        V1(
            LogLineV1 {
                date_time: 2024-01-16T16:03:04.440073230Z,
                method: "GET",
                url: "https://static.staging.crates.io/?1705420437",
                status: 403,
            },
        )
        "#);

        assert_eq!(
            output.date_time().to_string(),
            "2024-01-16 16:03:04.440073230 UTC"
        );
        assert_eq!(output.method(), "GET");
        assert_eq!(output.url(), "https://static.staging.crates.io/?1705420437");
        assert_eq!(output.status(), 403);

        match output {
            LogLine::V1(l) => {
                assert!(is_borrowed(&l.method));
                assert!(is_borrowed(&l.url));
            }
        }
    }

    #[allow(clippy::ptr_arg)]
    fn is_borrowed(s: &Cow<'_, str>) -> bool {
        match s {
            Cow::Borrowed(_) => true,
            Cow::Owned(_) => false,
        }
    }
}
