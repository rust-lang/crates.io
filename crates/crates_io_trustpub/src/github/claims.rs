use crate::github::GITHUB_ISSUER_URL;
use crate::github::workflows::extract_workflow_filename;
use chrono::serde::ts_seconds;
use chrono::{DateTime, Utc};
use jsonwebtoken::errors::{Error, ErrorKind};
use jsonwebtoken::{Algorithm, DecodingKey, Validation};

/// Claims extracted from a GitHub Actions OIDC token.
///
/// This struct is used to decode and validate the JWT token generated by
/// GitHub Actions. It contains the claims that are relevant for our "Trusted
/// Publishing" implementation.
///
/// See <https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token>.
#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct GitHubClaims {
    pub aud: String,
    #[serde(with = "ts_seconds")]
    pub iat: DateTime<Utc>,
    #[serde(with = "ts_seconds")]
    pub exp: DateTime<Utc>,
    pub jti: String,

    pub repository_owner_id: String,
    pub repository: String,
    pub workflow_ref: String,
    pub environment: Option<String>,
}

impl GitHubClaims {
    /// Decode and validate a JWT token, returning the relevant claims if valid.
    pub fn decode(token: &str, audience: &str, key: &DecodingKey) -> Result<Self, Error> {
        let validation = validation(audience);

        let claims: Self = jsonwebtoken::decode(token, key, &validation)?.claims;

        let leeway = chrono::TimeDelta::seconds(validation.leeway as i64);
        if claims.iat > Utc::now() + leeway {
            return Err(ErrorKind::ImmatureSignature.into());
        }

        Ok(claims)
    }

    /// Extract the workflow filename from the [`workflow_ref`](Self::workflow_ref)
    /// field or return `None` if the filename cannot be extracted.
    pub fn workflow_filename(&self) -> Option<&str> {
        extract_workflow_filename(&self.workflow_ref)
    }
}

fn validation(audience: &str) -> Validation {
    let mut validation = Validation::new(Algorithm::RS256);
    validation.required_spec_claims.insert("iss".into());
    validation.required_spec_claims.insert("exp".into());
    validation.required_spec_claims.insert("aud".into());
    validation.validate_exp = true;
    validation.validate_aud = true;
    validation.validate_nbf = true;
    validation.set_issuer(&[GITHUB_ISSUER_URL]);
    validation.set_audience(&[audience]);
    validation
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_keys::{DECODING_KEY, encode_for_testing};
    use insta::{assert_compact_debug_snapshot, assert_json_snapshot};
    use serde_json::json;
    use std::time::SystemTime;

    const AUDIENCE: &str = "crates.io";

    #[test]
    fn test_decode() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "sub": "repo:octo-org/octo-repo:environment:prod",
          "environment": "prod",
          "aud": AUDIENCE,
          "ref": "refs/heads/main",
          "sha": "example-sha",
          "repository": "octo-org/octo-repo",
          "repository_owner": "octo-org",
          "actor_id": "12",
          "repository_visibility": "private",
          "repository_id": "74",
          "repository_owner_id": "65",
          "run_id": "example-run-id",
          "run_number": "10",
          "run_attempt": "2",
          "runner_environment": "github-hosted",
          "actor": "octocat",
          "workflow": "example-workflow",
          "head_ref": "",
          "base_ref": "",
          "event_name": "workflow_dispatch",
          "ref_type": "branch",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "nbf": now,
          "exp": now + 30,
          "iat": now,
        }))?;

        let claims = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY)?;
        assert_json_snapshot!(claims, { ".iat" => "[datetime]", ".exp" => "[datetime]" }, @r#"
        {
          "aud": "crates.io",
          "iat": "[datetime]",
          "exp": "[datetime]",
          "jti": "example-id",
          "repository_owner_id": "65",
          "repository": "octo-org/octo-repo",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "environment": "prod"
        }
        "#);

        Ok(())
    }

    #[test]
    fn test_decode_minimal() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 30,
          "iat": now,
        }))?;

        let claims = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY)?;
        assert_json_snapshot!(claims, { ".iat" => "[datetime]", ".exp" => "[datetime]" }, @r#"
        {
          "aud": "crates.io",
          "iat": "[datetime]",
          "exp": "[datetime]",
          "jti": "example-id",
          "repository_owner_id": "65",
          "repository": "octo-org/octo-repo",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "environment": null
        }
        "#);

        Ok(())
    }

    #[test]
    fn test_decode_missing_jti() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 30,
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r#"Error(Json(Error("missing field `jti`", line: 1, column: 251)))"#);

        Ok(())
    }

    #[test]
    fn test_decode_wrong_audience() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": "somebody-else",
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 30,
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r"Error(InvalidAudience)");

        Ok(())
    }

    #[test]
    fn test_decode_multi_audience() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": [AUDIENCE, "somebody-else"],
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 30,
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r#"Error(Json(Error("invalid type: sequence, expected a string", line: 1, column: 7)))"#);

        Ok(())
    }
    #[test]
    fn test_decode_missing_repo() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 30,
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r#"Error(Json(Error("missing field `repository`", line: 1, column: 236)))"#);

        Ok(())
    }
    #[test]
    fn test_decode_missing_owner_id() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 30,
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r#"Error(Json(Error("missing field `repository_owner_id`", line: 1, column: 243)))"#);

        Ok(())
    }
    #[test]
    fn test_decode_missing_workflow() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 30,
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r#"Error(Json(Error("missing field `workflow_ref`", line: 1, column: 185)))"#);

        Ok(())
    }

    #[test]
    fn test_decode_missing_issuer() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "exp": now + 30,
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r#"Error(MissingRequiredClaim("iss"))"#);

        Ok(())
    }

    #[test]
    fn test_decode_wrong_issuer() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://gitlab.com",
          "exp": now + 30,
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @"Error(InvalidIssuer)");

        Ok(())
    }

    #[test]
    fn test_decode_missing_exp() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "iat": now,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r#"Error(Json(Error("missing field `exp`", line: 1, column: 253)))"#);

        Ok(())
    }

    #[test]
    fn test_decode_expired() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now - 3000,
          "iat": now - 6000,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @"Error(ExpiredSignature)");

        Ok(())
    }

    #[test]
    fn test_decode_missing_iat() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 30,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @r#"Error(Json(Error("missing field `iat`", line: 1, column: 253)))"#);

        Ok(())
    }

    #[test]
    fn test_decode_future_iat() -> anyhow::Result<()> {
        let now = SystemTime::UNIX_EPOCH.elapsed()?.as_secs();

        let jwt = encode_for_testing(&json!({
          "jti": "example-id",
          "aud": AUDIENCE,
          "repository": "octo-org/octo-repo",
          "repository_owner_id": "65",
          "workflow_ref": "octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main",
          "iss": "https://token.actions.githubusercontent.com",
          "exp": now + 300,
          "iat": now + 100,
        }))?;

        let error = GitHubClaims::decode(&jwt, AUDIENCE, &DECODING_KEY).unwrap_err();
        assert_compact_debug_snapshot!(error, @"Error(ImmatureSignature)");

        Ok(())
    }
}
