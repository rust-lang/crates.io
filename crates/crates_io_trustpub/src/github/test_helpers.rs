use crate::github::GITHUB_ISSUER_URL;
use crate::test_keys::encode_for_testing;
use bon::bon;
use serde_json::json;

pub const AUDIENCE: &str = "crates.io";

/// A struct representing all the claims in a GitHub Actions OIDC token.
///
/// This struct is used to create a JWT for testing purposes.
#[derive(Debug, serde::Serialize)]
pub struct FullGitHubClaims {
    pub iss: String,
    pub nbf: i64,
    pub exp: i64,
    pub iat: i64,
    pub jti: String,
    pub sub: String,
    pub aud: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub environment: Option<String>,
    #[serde(rename = "ref")]
    pub r#ref: String,
    pub sha: String,
    pub repository: String,
    pub repository_owner: String,
    pub actor_id: String,
    pub repository_visibility: String,
    pub repository_id: String,
    pub repository_owner_id: String,
    pub run_id: String,
    pub run_number: String,
    pub run_attempt: String,
    pub runner_environment: String,
    pub actor: String,
    pub workflow: String,
    pub head_ref: String,
    pub base_ref: String,
    pub event_name: String,
    pub ref_type: String,
    pub workflow_ref: String,
}

#[bon]
impl FullGitHubClaims {
    #[builder]
    pub fn new(
        owner_id: i32,
        owner_name: &str,
        repository_name: &str,
        workflow_filename: &str,
        environment: Option<&str>,
    ) -> Self {
        let now = chrono::Utc::now().timestamp();

        Self {
            iss: GITHUB_ISSUER_URL.into(),
            nbf: now,
            iat: now,
            exp: now + 30 * 60,
            jti: "example-id".into(),
            sub: format!("repo:{owner_name}/{repository_name}"),
            aud: AUDIENCE.into(),

            environment: environment.map(|s| s.into()),
            r#ref: "refs/heads/main".into(),
            sha: "example-sha".into(),
            repository: format!("{owner_name}/{repository_name}"),
            repository_owner: owner_name.into(),
            actor_id: "12".into(),
            repository_visibility: "private".into(),
            repository_id: "74".into(),
            repository_owner_id: owner_id.to_string(),
            run_id: "example-run-id".into(),
            run_number: "10".into(),
            run_attempt: "2".into(),
            runner_environment: "github-hosted".into(),
            actor: "octocat".into(),
            workflow: "example-workflow".into(),
            head_ref: "".into(),
            base_ref: "".into(),
            event_name: "workflow_dispatch".into(),
            ref_type: "branch".into(),
            workflow_ref: format!(
                "{owner_name}/{repository_name}/.github/workflows/{workflow_filename}@refs/heads/main"
            ),
        }
    }

    pub fn encoded(&self) -> anyhow::Result<String> {
        Ok(encode_for_testing(self)?)
    }

    pub fn as_exchange_body(&self) -> anyhow::Result<String> {
        let jwt = self.encoded()?;
        Ok(serde_json::to_string(&json!({ "jwt": jwt }))?)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use claims::assert_ok;
    use insta::assert_json_snapshot;

    #[test]
    fn test_github_claims() {
        let claims = FullGitHubClaims::builder()
            .owner_id(123)
            .owner_name("octocat")
            .repository_name("hello-world")
            .workflow_filename("ci.yml")
            .build();

        assert_json_snapshot!(claims, {
            ".nbf" => "[timestamp]",
            ".iat" => "[timestamp]",
            ".exp" => "[timestamp]",
        });

        let encoded = assert_ok!(claims.encoded());
        assert!(!encoded.is_empty());

        let exchange_body = assert_ok!(claims.as_exchange_body());
        assert!(exchange_body.contains(&encoded));
    }
}
