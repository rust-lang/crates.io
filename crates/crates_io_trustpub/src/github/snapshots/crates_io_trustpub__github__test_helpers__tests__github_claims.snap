---
source: crates/crates_io_trustpub/src/github/test_helpers.rs
expression: claims
---
{
  "iss": "https://token.actions.githubusercontent.com",
  "nbf": "[timestamp]",
  "exp": "[timestamp]",
  "iat": "[timestamp]",
  "jti": "example-id",
  "sub": "repo:octocat/hello-world",
  "aud": "crates.io",
  "ref": "refs/heads/main",
  "sha": "example-sha",
  "repository": "octocat/hello-world",
  "repository_owner": "octocat",
  "actor_id": "12",
  "repository_visibility": "private",
  "repository_id": "74",
  "repository_owner_id": "123",
  "run_id": "example-run-id",
  "run_number": "10",
  "run_attempt": "2",
  "runner_environment": "github-hosted",
  "actor": "octocat",
  "workflow": "example-workflow",
  "head_ref": "",
  "base_ref": "",
  "event_name": "workflow_dispatch",
  "ref_type": "branch",
  "workflow_ref": "octocat/hello-world/.github/workflows/ci.yml@refs/heads/main"
}
